ablateLibrary/domain/subDomain.cpp

```diff
@@ -153,3 +220,158 @@ void ablate::domain::SubDomain::ProjectFieldFunctions(const std::vector<std::sha
+        DMCreateGlobalVector(subDM, &subSolutionVec) >> checkError;
+
+        // Assign a name the same as the global
+        const char* vecName;
+        PetscObjectGetName((PetscObject)GetSolutionVector(), &vecName) >> checkError;
+        PetscObjectSetName((PetscObject)subSolutionVec, vecName) >> checkError;
+    }
+
+    CopyGlobalToSubVector(subDM, GetDM(), subSolutionVec, GetSolutionVector(), GetFields());
+
+    return subSolutionVec;
+}
+
+DM ablate::domain::SubDomain::GetSubAuxDM() {
+    // If there is no auxDM, there cannot be a sub Aux DM
+    if (!auxDM) {
+        return nullptr;
+    }
+
+    // If there is no label, just return the entire dm
+    if (!label) {
+        return GetAuxDM();
+    }
+
+    // If it is already created, return it
+    if (subAuxDM) {
+        return subAuxDM;
+    }
+
+    // Create a sub auxDM
+    /* MUST call DMGetCoordinateDM() in order to get p4est setup if present */
+    DM coordDM;
+    DMGetCoordinateDM(GetSubDM(), &coordDM) >> checkError;
+    DMClone(GetSubDM(), &subAuxDM) >> checkError;
+
+    // this is a hard coded "dmAux" that petsc looks for
+    DMSetCoordinateDM(subAuxDM, coordDM) >> checkError;
+
+    // Add all of the fields
+    // copy over all fields that were in the main dm
+    for (auto& fieldInfo : GetFields(FieldType::AUX)) {
+        auto petscField = GetPetscFieldObject(fieldInfo);
+        DMAddField(subAuxDM, NULL, petscField) >> checkError;
+    }
+
+    return subAuxDM;
+}
+
+Vec ablate::domain::SubDomain::GetSubAuxVector() {
+    // If there is no auxVector, return null
+    if (!auxVec) {
+        return nullptr;
+    }
+
+    // If there is no label, just return the entire solution vector
+    if (!label) {
+        return GetAuxVector();
+    }
+
+    if (!subAuxVec) {
+        DMCreateGlobalVector(GetSubAuxDM(), &subAuxVec) >> checkError;
+    }
+
+    CopyGlobalToSubVector(GetSubAuxDM(), GetAuxDM(), subAuxVec, GetAuxVector(), GetFields(FieldType::AUX), GetFields(FieldType::AUX), true);
+    return subAuxVec;
+}
+
+void ablate::domain::SubDomain::CopyGlobalToSubVector(DM sDM, DM gDM, Vec subVec, Vec globVec, const std::vector<Field>& subFields, const std::vector<Field>& gFields, bool localVector) {
+    /* Get the map from the subVec to global */
+    IS subpointIS;
+    const PetscInt* subpointIndices = NULL;
+    DMPlexGetSubpointIS(sDM, &subpointIS) >> checkError;
+    ISGetIndices(subpointIS, &subpointIndices) >> checkError;
+
+    // Get array access to the vec
+    const PetscScalar* globalVecArray;
+    PetscScalar* subVecArray;
+    VecGetArrayRead(globVec, &globalVecArray) >> checkError;
+    VecGetArray(subVec, &subVecArray) >> checkError;
+
+    // March over the global section
+    PetscSection section;
+    DMGetGlobalSection(sDM, &section) >> checkError;
+
+    PetscInt pStart, pEnd;
+    PetscSectionGetChart(section, &pStart, &pEnd) >> checkError;
+
+    // For each field in the subDM
+    for (std::size_t i = 0; i < subFields.size(); i++) {
```

Comment:
Note that this mostly reproduces the code in `DMCreateSubDM()` which returns an IS with these indices.

Reply:
Thanks @knepley!  Are there any support functions for moving data with the IS or should I iterate through it accessing array data?

Reply:
Yes, you call [VecISCopy](https://petsc.org/main/docs/manualpages/Vec/VecISCopy.html), (VecISSet)[https://petsc.org/main/docs/manualpages/Vec/VecISSet.html], [VecISAXPY](https://petsc.org/main/docs/manualpages/Vec/VecISAXPY.html)